### PR TITLE
consensus/pbft: replace gob with rlp

### PR DESCRIPTION
**Pull request by bailantaotao**
1. Replace gob with rlp. 
(Means that you do not have to implement ToPayload() using RequestContexter interface)
2. The data type of Msg has changed in Message struct. (inteface{} -> []byte, reason by 4)
3. A nil data is not supported by rlp encode.
4. A strong data type is needed for rlp decode.
```
type Proposal struct {
	Header         *ProposalHeader
	RequestContext RequestContexter
	Signatures     [][]byte
}

func (b *Proposal) DecodeRLP(s *rlp.Stream) error {
	var proposal struct {
		Header       *ProposalHeader
		**BlockContext *types.Block**
		Signatures   [][]byte
	}

	if err := s.Decode(&proposal); err != nil {
		return err
	}
	b.Header, b.RequestContext, b.Signatures = proposal.Header, proposal.BlockContext, proposal.Signatures
	return nil
}
```

### CODE REVIEW

File: consensus/pbft/backends/simple/engine.go [hunk 1]
```diff
@@ -301,6 +302,8 @@ func (sb *simpleBackend) Seal(chain consensus.ChainReader, block *types.Block, s
 		BlockContext: block,
 	})
 
+	// compensation for a few milliseconds of consensus runtime
```
Comment (bailantaotao) [inline consensus/pbft/backends/simple/engine.go]: todo: config delay time

File: consensus/pbft/core/types.go [hunk 2]
```diff
@@ -35,14 +35,43 @@ const (
...
+	Msg       []byte
 	Address   common.Address
 	Signature []byte
 }
 
+// ==============================================
+//
+// define the functions that needs to be provided for rlp Encoder/Decoder.
+
+// EncodeRLP serializes b into the Ethereum RLP View format.
```
Comment (yutelin) [inline consensus/pbft/core/types.go]: comment: what is b?

File: consensus/pbft/types.go [hunk 3]
```diff
@@ -33,43 +35,17 @@ type State struct {
 }
 
 // BlockContexter supports retrieving height and serialized block to be used during PBFT consensus.
-type BlockContexter interface {
+type RequestContexter interface {
 	// Number retrieves block height.
-	Height() *big.Int
+	Number() *big.Int
```
Comment (bailantaotao) [inline consensus/pbft/types.go]: Add Hash()

File: consensus/pbft/core/checkpoint.go [hunk 4]
```diff
@@ -25,17 +25,24 @@ import (
 func (c *core) sendCheckpoint(cp *pbft.Subject) {
 	logger := c.logger.New("state", c.state)
 	logger.Debug("sendCheckpoint")
+
+	newCp, err := Encode(cp)
+	if err != nil {
+		logger.Error("Failed to encode...")
```
Comment (yutelin) [inline consensus/pbft/core/checkpoint.go]: don't use three dots.

File: consensus/pbft/core/commit.go [hunk 5]
```diff
@@ -25,18 +25,24 @@ import (
 func (c *core) sendCommit() {
 	logger := c.logger.New("state", c.state)
 	logger.Debug("sendCommit")
+
+	subject, err := Encode(c.subject)
+	if err != nil {
+		logger.Error("Failed to encode...")
```
Comment (yutelin) [inline consensus/pbft/core/commit.go]: same here.

File: consensus/pbft/core/prepare.go [hunk 6]
```diff
@@ -25,18 +25,24 @@ import (
 func (c *core) sendPrepare() {
 	logger := c.logger.New("state", c.state)
 	logger.Debug("sendPrepare")
+
+	subject, err := Encode(c.subject)
+	if err != nil {
+		logger.Error("Failed to encode...")
```
Comment (yutelin) [inline consensus/pbft/core/prepare.go]: same here.

File: consensus/pbft/types.go [hunk 7]
```diff
@@ -69,6 +73,23 @@ func (b *BlockContext) Payload() []byte {
...
+func (b *BlockContext) DecodeRLP(s *rlp.Stream) error {
+	var proposal struct {
+		RawData []byte
+		Height *big.Int
+	}
+
+	if err := s.Decode(&proposal); err != nil {
+		return err
+	}
+	b.RawData, b.Height = b.RawData, proposal.Height
```
Comment (yutelin) [inline consensus/pbft/types.go]: proposal.RawData

File: consensus/pbft/backends/simple/backend.go [hunk 8]
```diff
@@ -124,10 +125,9 @@ func (sb *simpleBackend) Commit(proposal *pbft.Proposal) error {
 	// step1: update validator set from extra data of block
 	// step2: insert chain
 	block := &types.Block{}
-	err := rlp.DecodeBytes(proposal.BlockContext.Payload(), block)
-	if err != nil {
-		sb.logger.Warn("decode block error", "err", err)
-		return err
+	block, ok := proposal.BlockContext.(*types.Block)
+	if !ok {
+		return fmt.Errorf("failed to decode block...")
```
Comment (yutelin) [inline consensus/pbft/backends/simple/backend.go]: print useful message here.
Capitalize the first character.

File: consensus/pbft/backends/simple/backend.go [hunk 9]
```diff
@@ -178,10 +178,9 @@ func (sb *simpleBackend) EventMux() *event.TypeMux {
 func (sb *simpleBackend) Verify(proposal *pbft.Proposal) error {
 	// decode the proposal to block
 	block := &types.Block{}
-	err := rlp.DecodeBytes(proposal.BlockContext.Payload(), block)
-	if err != nil {
-		log.Warn("decode block error", "err", err)
-		return err
+	block, ok := proposal.BlockContext.(*types.Block)
+	if !ok {
+		return fmt.Errorf("failed to decode block...")
```
Comment (yutelin) [inline consensus/pbft/backends/simple/backend.go]: same here

File: consensus/pbft/types.go [hunk 10]
```diff
@@ -118,8 +199,29 @@ type NewView struct {
...
+	}
+
+	if err := s.Decode(&newView); err != nil {
+		return err
+	}
+	b.ViewNumber, b.VSet, b.XSet, b.Proposal = newView.ViewNumber, newView.VSet, newView.XSet, newView.Proposal
+	return nil
+}
+
+type Checkpoint struct {
```
Comment (tailingchen) [inline consensus/pbft/types.go]: remove it

File: consensus/pbft/backends/simple/backend.go [hunk 11]
```diff
@@ -124,10 +125,9 @@ func (sb *simpleBackend) Commit(proposal *pbft.Proposal) error {
 	// step1: update validator set from extra data of block
 	// step2: insert chain
 	block := &types.Block{}
-	err := rlp.DecodeBytes(proposal.BlockContext.Payload(), block)
-	if err != nil {
-		sb.logger.Warn("decode block error", "err", err)
-		return err
+	block, ok := proposal.RequestContext.(*types.Block)
+	if !ok {
+		return fmt.Errorf("failed to decode block...")
```
Comment (markya0616) [inline consensus/pbft/backends/simple/backend.go]: Suggest print log by sb.logger

File: consensus/pbft/backends/simple/backend.go [hunk 12]
```diff
@@ -178,10 +178,9 @@ func (sb *simpleBackend) EventMux() *event.TypeMux {
 func (sb *simpleBackend) Verify(proposal *pbft.Proposal) error {
 	// decode the proposal to block
 	block := &types.Block{}
-	err := rlp.DecodeBytes(proposal.BlockContext.Payload(), block)
-	if err != nil {
-		log.Warn("decode block error", "err", err)
-		return err
+	block, ok := proposal.RequestContext.(*types.Block)
+	if !ok {
+		return fmt.Errorf("failed to decode block...")
```
Comment (markya0616) [inline consensus/pbft/backends/simple/backend.go]: Suggest print log by sb.logger

File: consensus/pbft/backends/simple/engine.go [hunk 13]
```diff
@@ -296,15 +297,13 @@ func (sb *simpleBackend) Seal(chain consensus.ChainReader, block *types.Block, s
...
-		return nil, e
-	}
+	// step 2. post block into PBFT engine
 	go sb.EventMux().Post(pbft.RequestEvent{
-		BlockContext: pbft.NewBlockContext(b, block.Number()),
+		BlockContext: block,
 	})
 
+	// compensation for a few milliseconds of consensus runtime
+	<-time.After(100 * time.Millisecond)
```
Comment (markya0616) [inline consensus/pbft/backends/simple/engine.go]: Suggest to add delay before broadcast event

File: consensus/pbft/message_set_test.go [hunk 14]
```diff
@@ -22,10 +22,23 @@ import (
 	"testing"
 
 	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
```
Comment (markya0616) [inline consensus/pbft/message_set_test.go]: remove line

File: consensus/pbft/core/types.go (hunk 2 above)
Comment (bailantaotao) [inline consensus/pbft/core/types.go]: fixed

File: consensus/pbft/core/checkpoint.go (hunk 4 above)
Comment (bailantaotao) [inline consensus/pbft/core/checkpoint.go]: fixed in bce8dfc

File: consensus/pbft/core/commit.go (hunk 5 above)
Comment (bailantaotao) [inline consensus/pbft/core/commit.go]: fixed in bce8dfc

**Comment by bailantaotao** [comment]
all comments are fixed in bce8dfc

File: consensus/pbft/backends/simple/backend.go [hunk 15]
```diff
@@ -120,14 +121,15 @@ func (sb *simpleBackend) Broadcast(payload []byte) error {
...
 	block := &types.Block{}
-	err := rlp.DecodeBytes(proposal.BlockContext.Payload(), block)
-	if err != nil {
-		sb.logger.Warn("decode block error", "err", err)
-		return err
+	block, ok := proposal.RequestContext.(*types.Block)
+	if !ok {
+		errStr := "Failed to commit proposal since RequestContext cannot cast to *types.Block"
+		sb.logger.Error(errStr)
+		return fmt.Errorf(errStr)
```
Comment (markya0616) [inline consensus/pbft/backends/simple/backend.go]: errors.New? or predefined this error?

File: consensus/pbft/core/checkpoint.go [hunk 16]
```diff
@@ -28,7 +28,7 @@ func (c *core) sendCheckpoint(cp *pbft.Subject) {
 
 	newCp, err := Encode(cp)
 	if err != nil {
-		logger.Error("Failed to encode...")
+		logger.Error("Failed to encode", "subject", cp)
```
Comment (tailingchen) [inline consensus/pbft/core/checkpoint.go]: return function if error occurred?

**Pull request merged**